jaxley/io/graph.py

```diff
@@ -1,750 +1,874 @@
+        # l = 2*r ensures A_cylinder = 2*pi*r*l = 4*pi*r^2 = A_sphere.
+        # Here, we add another compartment such that there exists an `edge` between
+        # that new compartment and the single-point-soma of appropriate length.
+        swc_graph.nodes[1]["p"] = 0
+        for n in swc_graph.nodes:
+            swc_graph.nodes[n]["p"] += 1
+        swc_graph.add_node(0, **swc_graph.nodes[1])
+        swc_graph.nodes[0]["p"] = -1
+        swc_graph.add_edge(0, soma, l=2 * swc_graph.nodes[soma]["r"])
+        swc_graph = nx.relabel_nodes(swc_graph, {i: i + 1 for i in swc_graph.nodes})
+
+    undir_swc_graph = swc_graph.to_undirected()
+
+    # Handle somatic branchpoints. A somatic branchpoint is a branchpoint at which at
+    # least two connecting branches are somatic. In that case (and in the case of a
+    # single-point soma), non-somatic branches are assumed to start from their first
+    # traced point, not from the soma.
+    somatic_branchpoints = []
+    for node in soma_idxs:
+        # Somatic branchpoints are branchpoints (`.degree(node) > 2`)
+        if undir_swc_graph.degree(node) > 2:
+            num_somatic_neighbors = 0
+            # Somatic branchpoints must have at least 2 somatic neighbors.
+            for neighbor in undir_swc_graph.neighbors(node):
+                if undir_swc_graph.nodes[neighbor]["id"] == 1:
+                    num_somatic_neighbors += 1
+            if num_somatic_neighbors > 1:
+                somatic_branchpoints.append(node)
+    for soma in somatic_branchpoints:
+        # edges connecting nodes to soma are considered part of the soma -> l = 0.
+        for i, j in undir_swc_graph.edges(soma):
+            if undir_swc_graph.nodes[j]["id"] != 1:
+                undir_swc_graph.edges[i, j]["l"] = 0
+
+    branches, current_branch, all_type_ids = [], [], []
+
+    # Traverse the SWC graph and identify which nodes belong to one branch. This builds
+    # a list `branches: List` where each elements is a `np.array` of shape (N, 3).
+    # The `3` are `swc_parent, swc_node, length` of all SWC edges within a branch.
+    root = root if root else _find_root(undir_swc_graph)
+    undir_swc_graph.nodes[root]["type"] = "spurious"
+    for i, j in nx.dfs_edges(undir_swc_graph, root):
+        undir_swc_graph.nodes[j]["type"] = "spurious"
+        current_edge = (i, j)
+        current_len = undir_swc_graph.edges[current_edge]["l"]
+        current_type_id = undir_swc_graph.nodes[i]["id"]
+        current_branch += [(i, j, current_len)]
+        if _is_leaf(undir_swc_graph, j):
+            # If the SWC leaf node has a different type_id than the its SWC
+            # predecessor, then we explicitly must add both branches here.
+            if not _has_same_id(undir_swc_graph, i, j, relevant_type_ids):
+                # Add the branch that goes up until the last edge.
+                branches.append(current_branch[:-1])
+                all_type_ids.append(current_type_id)
```

Comment:
This can be just beneath the for loop, since its run in every if case if I am not mistaken